presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -0,0 +1,190 @@
+        List<Type> inputTypes = ImmutableList.of(type);
+        MethodHandle inputFunction;
+        MethodHandle outputFunction;
+        Class<? extends AccumulatorState> stateInterface = BigIntegerAndLongState.class;
+        AccumulatorStateSerializer<?> stateSerializer = new BigIntegerAndLongStateSerializer();
+
+        if (((DecimalType) type).isShort()) {
+            inputFunction = SHORT_DECIMAL_INPUT_FUNCTION;
+            outputFunction = SHORT_DECIMAL_OUTPUT_FUNCTION;
+        }
+        else {
+            inputFunction = LONG_DECIMAL_INPUT_FUNCTION;
+            outputFunction = LONG_DECIMAL_OUTPUT_FUNCTION;
+        }
+        inputFunction = inputFunction.bindTo(type);
+        outputFunction = outputFunction.bindTo(type);
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, type.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                createInputParameterMetadata(type),
+                inputFunction,
+                null,
+                null,
+                COMBINE_FUNCTION,
+                outputFunction,
+                stateInterface,
+                stateSerializer,
+                new BigIntegerAndLongStateFactory(),
+                type,
+                false);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, type, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(BLOCK_INPUT_CHANNEL, type), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void inputShortDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        accumulateValueInState(BigInteger.valueOf(type.getLong(block, position)), state);
+    }
+
+    public static void inputLongDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        accumulateValueInState(decodeUnscaledValue(type.getSlice(block, position)), state);
+    }
+
+    private static void accumulateValueInState(BigInteger value, BigIntegerAndLongState state)
+    {
+        initializeIfNeeded(state);
+        state.setBigInteger(state.getBigInteger().add(value));
+        state.setLong(state.getLong() + 1);
+    }
+
+    private static void initializeIfNeeded(BigIntegerAndLongState state)
+    {
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(BigInteger.valueOf(0));
+        }
+    }
+
+    public static void combine(BigIntegerAndLongState state, BigIntegerAndLongState otherState)
+    {
+        state.setLong(state.getLong() + otherState.getLong());
```

Comment:
I got myself really confused when I was trying to understand the correctness of this method. Your `deserialize` implementation suggests that `otherState.getBigInteger()` can be null. However, such case is not handled here and yet this code passes all tests.

I dug into it. It turns out there were some unintended guarantees provided by the implementation that calls `combine`. I removed those guarantees in #5637. I expect #5637 to be merged by the end of day today so that you can rebase this pull request on top of master. There will be test failures in your commits as a result. Please fix those failures.